File: src/derivation/parse-rule.ts
/**
 * Derivation Rule Parser
 *
 * Parses ORM derivation rules of the form:
 *   Consequent := Antecedent1 and Antecedent2 and ...
 *
 * Produces structured IR (DerivationRule) from free-text rules
 * using known noun names for tokenization (longest-first, same
 * strategy as src/claims/tokenize.ts).
 */

export interface RuleTriple {
  subject: string
  predicate: string
  object: string
  qualifier?: { predicate: string; object: string }
  comparison?: { op: '>' | '<' | '>=' | '<=' | '=' | '!='; value: number }
  literalValue?: string
}

export interface DerivationRule {
  text: string
  consequent: RuleTriple
  antecedents: RuleTriple[]
  kind: 'join' | 'comparison' | 'aggregate' | 'identity'
  aggregate?: { fn: string; noun: string }
}

type ComparisonOp = '>' | '<' | '>=' | '<=' | '=' | '!='

// ─── Helpers ────────────────────────────────────────────────────────

/**
 * Find all known nouns in text, returning them in order of appearance.
 * Uses longest-first matching to avoid partial matches (same as tokenize.ts).
 */
function findNouns(text: string, nouns: string[]): Array<{ name: string; start: number; end: number }> {
  if (nouns.length === 0) return []

  // Sort longest first so "Graph Schema" matches before "Graph"
  const sorted = [...nouns].sort((a, b) => b.length - a.length)

  const pattern = sorted.map((n) => n.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|')
  const regex = new RegExp('\\b(' + pattern + ')\\b', 'g')

  const found: Array<{ name: string; start: number; end: number }> = []
  let match: RegExpExecArray | null
  while ((match = regex.exec(text)) !== null) {
    found.push({ name: match[1], start: match.index, end: match.index + match[0].length })
  }
  return found
}

/**
 * Mask known noun names in text with null characters, so that " and " inside
 * noun names (e.g. "Supply and Demand") is not treated as a conjunct separator.
 * Returns { masked, unmask } where unmask() restores original noun names.
 */
function maskNouns(text: string, nouns: string[]): { masked: string; unmask: (s: string) => string } {
  if (nouns.length === 0) return { masked: text, unmask: (s) => s }

  const sorted = [...nouns].sort((a, b) => b.length - a.length)

  // Build replacement map: noun → placeholder with null chars
  const replacements: Array<{ original: string; placeholder: string }> = []
  let masked = text

  for (const noun of sorted) {
    if (!masked.includes(noun)) continue
    // Replace spaces in the noun with \0 to protect from splitting
    const placeholder = noun.replace(/ /g, '\0')
    if (placeholder === noun) continue // no spaces, no masking needed
    replacements.push({ original: noun, placeholder })
    // Use word-boundary-aware replacement
    const re = new RegExp('\\b' + noun.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') + '\\b', 'g')
    masked = masked.replace(re, placeholder)
  }

  const unmask = (s: string): string => {
    let result = s
    for (const { original, placeholder } of replacements) {
      result = result.split(placeholder).join(original)
    }
    return result
  }

  return { masked, unmask }
}

/**
 * Split RHS conjuncts on " and ", with noun-name masking to protect
 * noun names that contain "and".
 */
function splitConjuncts(rhs: string, nouns: string[]): string[] {
  const { masked, unmask } = maskNouns(rhs, nouns)
  const parts = masked.split(' and ')
  return parts.map((p) => unmask(p.trim()))
}

/**
 * Extract a comparison operator and value from the tail of a conjunct.
 * e.g. "Layer State has Valence > 0.3" → { op: '>', value: 0.3 } and cleaned text
 */
function extractComparison(text: string): { cleaned: string; comparison?: { op: ComparisonOp; value: number } } {
  const re = /\s*(>=|<=|!=|>|<|=)\s*(-?\d+(?:\.\d+)?)\s*$/
  const m = text.match(re)
  if (m) {
    return {
      cleaned: text.slice(0, m.index!).trim(),
      comparison: { op: m[1] as ComparisonOp, value: parseFloat(m[2]) },
    }
  }
  return { cleaned: text }
}

/**
 * Extract a literal value in single quotes from a conjunct.
 * e.g. "Layer State has Affect Region 'Excited'" → literalValue = 'Excited'
 */
function extractLiteral(text: string): { cleaned: string; literalValue?: string } {
  const re = /\s+'([^']+)'\s*$/
  const m = text.match(re)
  if (m) {
    return {
      cleaned: text.slice(0, m.index!).trim(),
      literalValue: m[1],
    }
  }
  return { cleaned: text }
}

/**
 * Parse a single clause (conjunct) into a RuleTriple.
 * Finds known nouns and extracts predicates between them.
 * Handles binary (2 nouns), ternary (3 nouns → qualifier), comparisons, and literals.
 */
function parseTriple(text: string, nouns: string[]): RuleTriple {
  // Strip leading "that " (used in identity rules)
  let cleaned = text.replace(/^that\s+/, '')

  // Extract literal and comparison
  const litResult = extractLiteral(cleaned)
  cleaned = litResult.cleaned
  const literalValue = litResult.literalValue

  const cmpResult = extractComparison(cleaned)
  cleaned = cmpResult.cleaned
  const comparison = cmpResult.comparison

  // Find nouns in cleaned text
  const found = findNouns(cleaned, nouns)

  if (found.length < 2) {
    // Fallback: return what we can
    return {
      subject: found.length > 0 ? found[0].name : '',
      predicate: cleaned,
      object: '',
      ...(literalValue !== undefined && { literalValue }),
      ...(comparison !== undefined && { comparison }),
    }
  }

  const subject = found[0].name
  const objectNoun = found.length >= 3 ? found[1] : found[found.length - 1]

  // Predicate: text between end of subject and start of object
  const subjectEnd = found[0].end
  const objectStart = objectNoun.start
  const predicate = cleaned.slice(subjectEnd, objectStart).trim()

  const triple: RuleTriple = {
    subject,
    predicate,
    object: objectNoun.name,
    ...(literalValue !== undefined && { literalValue }),
    ...(comparison !== undefined && { comparison }),
  }

  // Ternary: 3 nouns → qualifier (e.g. "Graph uses Resource for Role")
  if (found.length >= 3) {
    const qualifierNoun = found[found.length - 1]
    const qualifierPredicateStart = objectNoun.end
    const qualifierPredicateEnd = qualifierNoun.start
    const qualifierPredicate = cleaned.slice(qualifierPredicateStart, qualifierPredicateEnd).trim()
    triple.qualifier = {
      predicate: qualifierPredicate,
      object: qualifierNoun.name,
    }
  }

  return triple
}

// ─── Main Parser ────────────────────────────────────────────────────

export function parseRule(text: string, nouns: string[]): DerivationRule {
  // Strip trailing period
  const cleaned = text.replace(/\.\s*$/, '')

  // Split on :=
  const splitIdx = cleaned.indexOf(':=')
  if (splitIdx === -1) {
    throw new Error(`Derivation rule must contain ':=': ${text}`)
  }
  const lhs = cleaned.slice(0, splitIdx).trim()
  const rhs = cleaned.slice(splitIdx + 2).trim()

  // Parse consequent
  const consequent = parseTriple(lhs, nouns)

  // ── Detect rule kind ──

  // Identity: RHS contains "the same"
  const identityMatch = rhs.match(/\bthe same\b/)
  if (identityMatch) {
    const antecedent = parseIdentityTriple(rhs, nouns)
    return {
      text,
      consequent,
      antecedents: [antecedent],
      kind: 'identity',
    }
  }

  // Aggregate: RHS starts with aggregate function (count/sum/avg/min/max) of Noun where ...
  const aggregateMatch = rhs.match(/^(count|sum|avg|min|max)\s+of\s+(\w[\w\s]*?)\s+where\s+(.+)$/i)
  if (aggregateMatch) {
    const fn = aggregateMatch[1].toLowerCase()
    const aggNounText = aggregateMatch[2].trim()
    const whereClause = aggregateMatch[3].trim()

    // Resolve the aggregate noun against known nouns
    const aggNounMatches = findNouns(aggNounText, nouns)
    const aggNoun = aggNounMatches.length > 0 ? aggNounMatches[0].name : aggNounText

    // Parse where-clause conjuncts
    const conjuncts = splitConjuncts(whereClause, nouns)
    const antecedents = conjuncts.map((c) => parseTriple(c, nouns))

    return {
      text,
      consequent,
      antecedents,
      kind: 'aggregate',
      aggregate: { fn, noun: aggNoun },
    }
  }

  // Default: split RHS into conjuncts
  const conjuncts = splitConjuncts(rhs, nouns)
  const antecedents = conjuncts.map((c) => parseTriple(c, nouns))

  // Classify: if any antecedent has a comparison → 'comparison', else → 'join'
  const hasComparison = antecedents.some((a) => a.comparison !== undefined)
  const kind = hasComparison ? 'comparison' : 'join'

  return {
    text,
    consequent,
    antecedents,
    kind,
  }
}

/**
 * Parse an identity-form RHS like "that Domain is the same Domain".
 * The predicate is "is the same" and we find the noun on each side.
 */
function parseIdentityTriple(text: string, nouns: string[]): RuleTriple {
  // Strip leading "that "
  let cleaned = text.replace(/^that\s+/, '')

  // Find "is the same" and split around it
  const sameIdx = cleaned.indexOf('is the same')
  if (sameIdx === -1) {
    // Fallback: parse as normal triple
    return parseTriple(text, nouns)
  }

  const beforeSame = cleaned.slice(0, sameIdx).trim()
  const afterSame = cleaned.slice(sameIdx + 'is the same'.length).trim()

  // Find subject noun in beforeSame
  const subjectNouns = findNouns(beforeSame, nouns)
  const subject = subjectNouns.length > 0 ? subjectNouns[0].name : beforeSame

  // Find object noun in afterSame
  const objectNouns = findNouns(afterSame, nouns)
  const object = objectNouns.length > 0 ? objectNouns[0].name : afterSame

  return {
    subject,
    predicate: 'is the same',
    object,
  }
}

File: crates/fol-engine/src/parse_rule.rs
// crates/fol-engine/src/parse_rule.rs
//
// Derivation Rule Parser — parses ORM derivation rules of the form:
//   Consequent := Antecedent1 and Antecedent2 and ...
//
// Produces structured triples from free-text rules using known noun names.

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RuleTriple {
    pub subject: String,
    pub predicate: String,
    pub object: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub qualifier: Option<Qualifier>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comparison: Option<Comparison>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub literal_value: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Qualifier {
    pub predicate: String,
    pub object: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct Comparison {
    pub op: String,
    pub value: f64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DerivationRule {
    pub text: String,
    pub consequent: RuleTriple,
    pub antecedents: Vec<RuleTriple>,
    pub kind: String, // "join", "comparison", "aggregate", "identity"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub aggregate: Option<Aggregate>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Aggregate {
    #[serde(rename = "fn")]
    pub func: String,
    pub noun: String,
}

struct NounMatch {
    name: String,
    start: usize,
    end: usize,
}

fn find_nouns(text: &str, nouns: &[String]) -> Vec<NounMatch> {
    let mut sorted: Vec<&String> = nouns.iter().collect();
    sorted.sort_by(|a, b| b.len().cmp(&a.len()));

    let mut found = Vec::new();
    let lower = text.to_lowercase();

    for noun in sorted {
        let noun_lower = noun.to_lowercase();
        if let Some(start) = lower.find(&noun_lower) {
            // Check word boundaries
            let before_ok = start == 0 || !text.as_bytes()[start - 1].is_ascii_alphanumeric();
            let end = start + noun.len();
            let after_ok = end >= text.len() || !text.as_bytes()[end].is_ascii_alphanumeric();
            if before_ok && after_ok {
                found.push(NounMatch { name: noun.clone(), start, end });
            }
        }
    }

    found.sort_by_key(|m| m.start);
    found
}

fn split_conjuncts(rhs: &str, nouns: &[String]) -> Vec<String> {
    // Mask noun names containing "and" to avoid splitting on them
    let mut masked = rhs.to_string();
    let mut sorted: Vec<&String> = nouns.iter().filter(|n| n.contains(" and ") || n.contains(" And ")).collect();
    sorted.sort_by(|a, b| b.len().cmp(&a.len()));

    let mut replacements = Vec::new();
    for noun in sorted {
        let placeholder = noun.replace(' ', "\0");
        if masked.contains(noun.as_str()) {
            masked = masked.replace(noun.as_str(), &placeholder);
            replacements.push((noun.clone(), placeholder));
        }
    }

    let parts: Vec<String> = masked.split(" and ").map(|s| {
        let mut result = s.trim().to_string();
        for (original, placeholder) in &replacements {
            result = result.replace(placeholder, original);
        }
        result
    }).collect();

    parts
}

fn extract_comparison(text: &str) -> (String, Option<Comparison>) {
    let re = regex::Regex::new(r"\s*(>=|<=|!=|>|<|=)\s*(-?\d+(?:\.\d+)?)\s*$").unwrap();
    if let Some(caps) = re.captures(text) {
        let m = caps.get(0).unwrap();
        let cleaned = text[..m.start()].trim().to_string();
        let op = caps[1].to_string();
        let value: f64 = caps[2].parse().unwrap_or(0.0);
        return (cleaned, Some(Comparison { op, value }));
    }
    (text.to_string(), None)
}

fn extract_literal(text: &str) -> (String, Option<String>) {
    if let Some(idx) = text.rfind('\'') {
        if let Some(start) = text[..idx].rfind('\'') {
            let value = text[start + 1..idx].to_string();
            let cleaned = text[..start].trim().to_string();
            return (cleaned, Some(value));
        }
    }
    (text.to_string(), None)
}

fn parse_triple(text: &str, nouns: &[String]) -> RuleTriple {
    let cleaned = text.strip_prefix("that ").unwrap_or(text);

    let (cleaned, literal_value) = extract_literal(cleaned);
    let (cleaned, comparison) = extract_comparison(&cleaned);

    let found = find_nouns(&cleaned, nouns);

    if found.len() < 2 {
        return RuleTriple {
            subject: found.first().map(|m| m.name.clone()).unwrap_or_default(),
            predicate: cleaned,
            object: String::new(),
            qualifier: None,
            comparison,
            literal_value,
        };
    }

    let subject = found[0].name.clone();
    let object_match = if found.len() >= 3 { &found[1] } else { found.last().unwrap() };
    let predicate = cleaned[found[0].end..object_match.start].trim().to_string();

    let mut triple = RuleTriple {
        subject,
        predicate,
        object: object_match.name.clone(),
        qualifier: None,
        comparison,
        literal_value,
    };

    if found.len() >= 3 {
        let qual = found.last().unwrap();
        let qual_pred = cleaned[object_match.end..qual.start].trim().to_string();
        triple.qualifier = Some(Qualifier { predicate: qual_pred, object: qual.name.clone() });
    }

    triple
}

pub fn parse_rule(text: &str, nouns: &[String]) -> Result<DerivationRule, String> {
    let cleaned = text.trim_end_matches('.');

    let split_idx = match cleaned.find(":=") {
        Some(idx) => idx,
        None => return Err(format!("Derivation rule must contain ':=': {}", text)),
    };

    let lhs = cleaned[..split_idx].trim();
    let rhs = cleaned[split_idx + 2..].trim();

    let consequent = parse_triple(lhs, nouns);

    // Identity: "the same"
    if rhs.contains("the same") {
        let antecedent = parse_triple(rhs, nouns);
        return Ok(DerivationRule {
            text: text.to_string(), consequent, antecedents: vec![antecedent], kind: "identity".to_string(), aggregate: None,
        });
    }

    // Aggregate: "count of X where ..."
    let agg_re = regex::Regex::new(r"(?i)^(count|sum|avg|min|max)\s+of\s+(.+?)\s+where\s+(.+)$").unwrap();
    if let Some(caps) = agg_re.captures(rhs) {
        let func = caps[1].to_lowercase();
        let agg_noun_text = caps[2].trim();
        let where_clause = caps[3].trim();

        let agg_nouns = find_nouns(agg_noun_text, nouns);
        let agg_noun = agg_nouns.first().map(|m| m.name.clone()).unwrap_or_else(|| agg_noun_text.to_string());

        let conjuncts = split_conjuncts(where_clause, nouns);
        let antecedents: Vec<RuleTriple> = conjuncts.iter().map(|c| parse_triple(c, nouns)).collect();

        return Ok(DerivationRule {
            text: text.to_string(), consequent, antecedents, kind: "aggregate".to_string(),
            aggregate: Some(Aggregate { func, noun: agg_noun }),
        });
    }

    // Default: split on "and"
    let conjuncts = split_conjuncts(rhs, nouns);
    let antecedents: Vec<RuleTriple> = conjuncts.iter().map(|c| parse_triple(c, nouns)).collect();
    let has_comparison = antecedents.iter().any(|a| a.comparison.is_some());
    let kind = if has_comparison { "comparison" } else { "join" };

    Ok(DerivationRule {
        text: text.to_string(), consequent, antecedents, kind: kind.to_string(), aggregate: None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_join_rule() {
        let nouns = vec!["Constraint".to_string(), "Modality Type".to_string()];
        let rule = parse_rule("Constraint is semantic := Constraint has modality of Modality Type 'Deontic'.", &nouns).unwrap();
        assert_eq!(rule.kind, "join");
        assert_eq!(rule.consequent.subject, "Constraint");
        assert_eq!(rule.antecedents[0].literal_value.as_deref(), Some("Deontic"));
    }

    #[test]
    fn comparison_rule() {
        let nouns = vec!["Layer State".to_string(), "Arousal".to_string()];
        let rule = parse_rule("Layer State is alarmed := Layer State has Arousal > 0.8.", &nouns).unwrap();
        assert_eq!(rule.kind, "comparison");
        assert!(rule.antecedents[0].comparison.is_some());
        assert_eq!(rule.antecedents[0].comparison.as_ref().unwrap().op, ">");
    }

    #[test]
    fn missing_delimiter_errors() {
        let result = parse_rule("not a rule", &[]);
        assert!(result.is_err());
    }
}
